### PR TITLE
Add a virtual ShouldRetry method to the RetryPolicy for customization.

### PR DESCRIPTION
**Pull request by ahsonkhan**
Fixes https://github.com/Azure/azure-sdk-for-cpp/issues/5515

Related discussion:
https://github.com/Azure/azure-sdk-for-cpp/pull/5530#issuecomment-2076038171

Given this is not meant for end customers to modify, I did not add it to the changelog to avoid highlighting it.

cc @Jinming-Hu 

Here is how I would expect the storage use case to be implemented. The `StorageRetryPolicy` would inherit from `RetryPolicy`, and override the `ShouldRetry` method:
```C++
  bool ShouldRetry(std::unique_ptr<RawResponse> const& response, RetryOptions const& retryOptions)
      const override
  {
    if (response == nullptr)
    {
      return true;
    }

    if (static_cast<std::underlying_type_t<Azure::Core::Http::HttpStatusCode>>(
            response->GetStatusCode())
        >= 400)
    {
      const auto& headers = response->GetHeaders();
      auto ite = headers.find("x-ms-copy-source-status-code");
      if (ite != headers.end())
      {
        const auto innerStatusCodeInt = std::stoi(ite->second);
        const auto innerStatusCode
            = static_cast<Azure::Core::Http::HttpStatusCode>(innerStatusCodeInt);

        const bool shouldRetry
            = retryOptions.StatusCodes.find(innerStatusCode) != retryOptions.StatusCodes.end();

        if (Azure::Core::Diagnostics::_internal::Log::ShouldWrite(
                Azure::Core::Diagnostics::Logger::Level::Informational))
        {
          Azure::Core::Diagnostics::_internal::Log::Write(
              Azure::Core::Diagnostics::Logger::Level::Informational,
              std::string("x-ms-copy-source-status-code ") + std::to_string(innerStatusCodeInt)
                  + (shouldRetry ? " will be retried" : " won't be retried"));
        }
        if (shouldRetry)
        {
          return true;
        }
      }
    }
    return false;
  }
```

Follow-up improvement to remove such tight coupling in existing tests: https://github.com/Azure/azure-sdk-for-cpp/issues/5585

### CODE REVIEW

File: sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp:375]: These two functions are virtual just for testing, right?

**Review by Jinming-Hu — APPROVED** [review]

File: sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp:375]: Yes. I actually want to make them non-virtual/private and update the tests, at some point.

**Comment by Jinming-Hu** [comment]
@ahsonkhan can we merge this PR as soon as possible? This is blocking our STG94 work.

**Comment by RickWinter** [comment]
@LarryOsterman can you please review as well.

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (LarryOsterman) [inline sdk/core/azure-core/src/http/retry_policy.cpp:125]: Are you planning on documenting this function? If that's the case, keeping these named parameters makes sense, but if you're not planning on documenting this function, I suggest removing the named parameters, which avoids the ugly cast to `void`.

**Review by LarryOsterman — APPROVED** [review]

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/retry_policy.cpp:125]: Agreed. It's already documented in the public header file. I won't be documenting it here in the source, so will update accordingly.

**Comment by ahsonkhan** [comment]
> can we merge this PR as soon as possible? This is blocking our STG94 work.

We are waiting for this month's release cycle for May before merging this. So, I expect us to get this checked-in by eow. My current thinking is that we beta this as part of Core's release in June and GA it in July to align with when storage needs it.

**Comment by Jinming-Hu** [comment]
@ahsonkhan core release is done. Can we merge this now?

**Pull request merged**
